### ISSUE TITLE
Can not connect dirtyforms with requirejs

### ISSUE THREAD

**Issue by ar4ers** [snikch/jquery.dirtyforms#82]
https://github.com/snikch/jquery.dirtyforms/blob/master/jquery.dirtyforms.js#L12
At this row function `factory` called without its `window` and `document` parameters, which leads to application crash.

**Comment by NightOwl888** [comment]
Thanks.

I am not exactly sure this is correct, nor do I know how to test it, so if you could test out the change in commit [d71ee0f89d6eeaef91835d0b88141158b3e1e3bb](https://github.com/snikch/jquery.dirtyforms/commit/d71ee0f89d6eeaef91835d0b88141158b3e1e3bb), it would be most helpful. Once I hear back it is working, I will create another beta release. I followed the AMD page [here](http://requirejs.org/docs/whyamd.html#definition) for this fix.

**Comment by ar4ers** [comment]
I tested. It doesn't work, cose requirejs is trying to load modules with names `window.js` and `document.js` instead of global `window` and `document` objects.
I don't know how exactly it can be done using `requirejs` module definition, but it can easily be tested.
Here some code for you:

```javascript
        <script src="js/require.js"></script>
        <script>
            requirejs.config({
                baseUrl: "js",
                paths: {
                    jquery: 'libs/jquery-2.1.4',
                    dirtyforms: "libs/jquery.dirtyforms"
                }
            });
            requirejs(["dirtyforms"],
                    function () {
                        console.log('loaded');
                    });
        </script>
```
I assume that I have `js/libs/` folder, containing `jquery` and `dirtyforms`, relative to this script.

**Comment by NightOwl888** [comment]
Thanks.

I will experiment with the code to test later.

In the meantime, could you let me know if it works when using the code from the [Making your jQuery plugin work better with npm tools](http://blog.npmjs.org/post/112712169830/making-your-jquery-plugin-work-better-with-npm) article?

```javascript
(function (factory) {
  if(typeof module === "object" && typeof module.exports === "object") {
    factory(require("jquery"), window, document);
  } else {
    factory(jQuery, window, document);
  }
}(function($, window, document, undefined) {
  // code for attaching tipso to whatever jQuery was passed in goes here.
}));
```

I have attempted to merge that code with the code from the [UMD for jQuery](https://github.com/umdjs/umd/blob/master/jqueryPluginCommonjs.js) source code, but being that I am not familiar with these environments, I am not sure exactly how to do the merge. But it occurs to me that the merge may not be required at all and the NPM page probably does everything that is required by requirejs.

**Comment by ar4ers** [comment]
I'm also not very familiar with this technologies(not my main interest), but I tried to merge you suggestion and it works like magic:

```javascript
(function (factory) {
    if (typeof define === 'function' && define.amd) {
        // AMD. Register as an anonymous module.
        factory(require("jquery"), window, document);
    } else if (typeof exports === 'object') {
        // Node/CommonJS
        module.exports = factory(require('jquery'), window, document);
    } else {
        // Browser globals
        factory(jQuery, window, document);
    }
}(function ($, window, document, undefined) {}));
```

Can't say if it's best way to solve your configuration, but it works :)

**Comment by NightOwl888** [comment]
Unfortunately, that would defeat the purpose. The whole idea is that you can inject the global variables from somewhere else so you can mock them if needed.

I found an example of how this is used in [jQuery](https://github.com/jquery/jquery/blob/master/src/var/document.js). You just need to create a file to wrap each global variable.

document.js
==

```js
define(function() { 
	return window.document; 
}); 
```

window.js
==

```js
define(function() { 
	return window; 
}); 
```

I should probably put these into a subdirectory and ship them inside of the packages so everything will work in production and on the CDN. But please try adding these files first to see if that works.

**Comment by ar4ers** [comment]
I also found this solution for the mocking purposes, but if it will be required to include this two additional files in production, it will complicate plugin and make its usage less convenient.

**Comment by NightOwl888** [comment]
Point taken.

It appears that with jQuery there is a build step that allows usage with test environments and then these pieces are stripped out for production use. The entire jQuery closure in the code download looks like this.

```js
(function( window, undefined ) {
    var document = window..document;
```

The raw source varies per module, here is the one from `core/init`.

```js
// Initialize a jQuery object
define([
	"../core",
	"../var/document",
	"./var/rsingleTag",
	"../traversing/findFilter"
], function( jQuery, document, rsingleTag ) {
```

As for what or how those pieces are stripped out, I haven't yet investigated.

BTW - I was unable to get it working with the code you provided (but I did manage to get jquery to load).

I am starting to wonder if it is really worth the extra effort to keep this UMD support in place.

**Comment by ar4ers** [comment]
Can't say :) You decide.

**Comment by NightOwl888** [comment]
Here is what I am thinking. Since NPM is supposed to be a backend development tool and the article I followed specifically said it was the NPM environment that should be configured that way, my thought is to keep the closure (almost) the way it is in the repo and for NPM, but to strip it out for Bower, NuGet, and CDN and replace it with the same closure jQuery is using in those environments.

The one change I am thinking to do to the closure is to remove the document element there and define it as a plugin-level variable the same way jQuery did. Its one less thing to futz with if you happen to be setting it up with requirejs/NPM.

So, basically those dependencies will be required if you use NPM or pull the source from the repo, but not if you use another package delivery method. Since I don't know whether you are using NPM, does that sound fair to you?

**Comment by NightOwl888** [comment]
@ar4ers - I would appreciate your feedback on how this works for you with RequireJS. As long as you are not using NPM the IIFE is now:

```js
(function($, window, document, undefined) {
    /// implementation...
})(jQuery, window, document);
```

So you don't have to pass it anything. 

If using NPM in conjunction with RequireJS (APM), then you need to add the files for `document` and `window`.

document.js
--

```js
define(function() { 
	return window.document; 
}); 
```

window.js
--

```js
define(function() { 
	return window; 
}); 
```

**Comment by LanmanGroup** [comment]
Same issue with webpack.
```
These dependencies were not found:

* document in ./node_modules/jquery.dirtyforms/jquery.dirtyforms.min.js
* window in ./node_modules/jquery.dirtyforms/jquery.dirtyforms.min.js
```
Already tried:

```js
require('./document');
require('./window');
require('jquery.dirtyforms');
```
